packages/workflow/src/workflow.ts

```diff
@@ -847,9 +847,20 @@ export function condition(fn: () => boolean, timeout: number | string): Promise<
  */
 export function condition(fn: () => boolean): Promise<void>;
 
-export function condition(fn: () => boolean, timeout?: number | string): Promise<void | boolean> {
+export async function condition(fn: () => boolean, timeout?: number | string): Promise<void | boolean> {
   if (timeout) {
-    return Promise.race([sleep(timeout).then(() => false), conditionInner(fn).then(() => true)]);
+    return await CancellationScope.cancellable(async () => {
+      return await Promise.race([
+        sleep(timeout).then(() => false),
+        conditionInner(fn).then(() => {
+          // Use patched to avoid breaking existing histories using this `condition`
+          if (patched('__temporal-internal-condition-cancels-timer')) {
+            CancellationScope.current().cancel();
```

Comment:
Forgive my lack of knowledge of the API, but since this is in a different part of the promise as a result of user code in `fn`, any concern that `CancellationScope.current()` may represent something besides the outer scope? I guess this may not be a concern because we don't allow async condition functions?

I wonder if we want to have `const cancelScope = CancellationScope.current()` as the first line in `cancellable` and change this to `cancelScope.cancel()` just to be safe.

Reply:
The scope is guaranteed to be the one we created because even if `fn` created a new scope it would be exited by the time `then` is called.
Now that I think of it, we should probably cancel the scope in case the timer resolves before the condition to clean up resources created by it.